modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ManagedLocalResult.java

```diff
@@ -43,71 +86,568 @@
+    @Override public int getRowId() {
+        return rowId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isAfterLast() {
+        return rowId >= rowCount;
+    }
+
+    /**
+     * @param values Values.
+     */
+    private void cloneLobs(Value[] values) {
+        for (int i = 0; i < values.length; i++) {
+            Value v = values[i];
+            Value v2 = v.copyToResult();
+            if (v2 != v) {
+                containsLobs = true;
+                session.addTemporaryLob(v2);
+                values[i] = v2;
+            }
+        }
+    }
+
+    /**
+     * @param values row.
+     * @return Row,
+     */
+    private ValueRow getDistinctRow(Value[] values) {
+        if (distinctIndexes != null) {
+            int cnt = distinctIndexes.length;
+            Value[] newValues = new Value[cnt];
+            for (int i = 0; i < cnt; i++) {
+                newValues[i] = values[distinctIndexes[i]];
+            }
+            values = newValues;
+        }
+        else if (values.length > visibleColumnCount) {
+            values = Arrays.copyOf(values, visibleColumnCount);
+        }
+        return ValueRow.get(values);
+    }
+
+    private void createExternalResult() {
+        assert ctx != null;
+
+        external = distinct || distinctIndexes != null || sort != null ?
+            new SortedExternalResult(ctx, session, distinct, distinctIndexes, visibleColumnCount, sort, memTracker,
+                rowCount) : new PlainExternalResult(ctx, memTracker, session);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addRow(Value[] values) {
+        cloneLobs(values);
+        if (isAnyDistinct()) {
+            if (distinctRows != null) {
+                ValueRow array = getDistinctRow(values);
+                Value[] previous = distinctRows.get(array);
+                if (previous == null || sort != null && sort.compare(previous, values) > 0) {
+                    distinctRows.put(array, values);
+                }
+                rowCount = distinctRows.size();
+                if (!hasAvailableMemory(array, previous, values)) {
+                    addRowsToDisk();
+
+                    distinctRows = null;
+                }
+            } else {
+                rowCount = external.addRow(values);
+            }
+        } else {
+            rowCount++;
+            if (external == null) {
+                rows.add(values);
+                if (!hasAvailableMemory(null, null, values)) {
+                    addRowsToDisk();
+                }
+            }
+            else
+                external.addRow(values);
+        }
+    }
+
+    /**
+     * Adds rows to disk.
+     */
+    private void addRowsToDisk() {
+        if (external == null) {
+            createExternalResult();
+        }
+
+        // We need to release memory before adding to external result to prevent immediate rows spilling.
+        if (memTracker != null)
+            memTracker.released(memReserved);
```

Comment:
Let's clear rows\distinctRows fields before memTracker.release().

Reply:
> Is it possible to cache value in some MemoryManager field and pass copy actual to context?
This will allow to change value between node starts and avoid unnecessary synchronous calls.

We will rework entire configuration for quotas and offloading. See https://ggsystems.atlassian.net/browse/GG-25920. I'm not sure it should be configured with the system property. We will change it to static or metastore value.

> Let's clear rows\distinctRows fields before memTracker.release().

I made a fix. Waiting for TC now.